src/AsyncProducers.cpp

```diff
@@ -569,11 +569,67 @@ class InitializeSemaphores : public IRMutator {
     }
 };
 
+// A class to support stmt_uses_vars queries that repeatedly hit the same
+// sub-stmts. Used to support TightenProducerConsumerNodes below.
+class CachingStmtUsesVars : public IRMutator {
+    const Scope<> &query;
+    bool found_use = false;
+    std::map<Stmt, bool> cache;
```

Comment:
Why not `std::set<>` instead?

Reply:
Also, just fer grins, would `unordered_map` or `unordered_set` be any better here? (Surely the order doesn't matter)

Reply:
There are three possible states: 
- I've seen this stmt before and it doesn't contain the vars (map contains false)
- I've seen this stmt before and it does contain the vars (map contains true)
- I haven't seen this stmt before and it needs to be analyzed (not in the map)

Reply:
I generally avoid unordered_set and unordered_map because every time I've benchmarked them within Halide they haven't been enough faster to outweigh a really annoying property that I've been burned by: If you have a bug where you accidentally depend on the order of the keys, it can fail on some machines but not others depending on the standard library used, which makes debugging it really annoying.